core/lib/spree/core/error_reporter/base.rb

```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+module Spree
+  module Core
+    module ErrorReporter
+      ##
+      # Abstract class only meant to be inherited from
+      #
+      # @example Create a custom error handler
+      #   class CustomErrorReporter < Spree::Core::ErrorReporter::Base
+      #     # Output to as debug by default
+      #     def report(error, severity, metadata)
+      #       Rollbar.send(severity, error, metadata)
+      #     end
+      #   end
+      #
+      #   Spree::Core::ErrorReporter.add_reporter(CustomErrorReporter)
+      #
+      class Base
+        class << self
+          ##
+          # Generic error handling
+          # @param error [StandardError] The error you want to handle.
+          # @param serverity [Symbol, String] The severity (i.e. debug, info, warn, error, fatal)
+          # @param metadata [Hash] Any metadata we want to pass in
+          #
+          def report(_error, _severity = :error, _metadata = {})
+            raise NoMethodError
```

Comment:
Might be better to have it `raise NotImplementedError`

Reply:
As a developer, if I see an error that says `NotImplementedError`, that really tells me something :p

But if you see `NoMethodError` I would check to see if I spelled the method right. `NoMethodError` covers a wider array of reasons why it gets raised.

If I see someone do: `rescue NoMethodError` I would wonder why. Are they getting `nil` and `rescue`'ing if it ain't there? So many questions with `NoMethodError`.

The purpose of `NotImplementedError` was to communicate what needs to be implemented. You've spelled your method right and if you rescue it, it's obvious to give a default response when a method isn't implemented or to raise a new default error that is better formatted to the appropriate context.

Reply:
Ah, just read the link. This puts a spin on things: http://ruby-doc.org/core-2.5.1/NotImplementedError.html

> Raised when a feature is not implemented on the current platform. For example, methods depending on the fsync or fork system calls may raise this exception if the underlying operating system or Ruby runtime does not support them.
>
> Note that if fork raises a NotImplementedError, then respond_to?(:fork) returns false.